### PR TITLE
DOC: Adding a high-level description of our software design process

### PR DESCRIPTION
**Pull request by ant0nsc**

### CODE REVIEW

**Review by javier-alvarez — COMMENTED** [review]
Looks good but check comments

File: docs/source/software_design_overview.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+Both of those package management systems maintain strict versioning: Once a version of a package is published, it
+cannot be modified in place. Rather, a new version needs to be released.
+
+Our training and deployment code uses Conda environment files that specify an explicit version of a dependent package to
+use (for example, `lightning_bolts==0.4.0`). The Conda environment files are also version controlled in GitHub. Any
+change to a version of a 3rd party library will need to be carried out via the same change management process as a code
+change, with Pull Request, review, and all tests passing.
+
+The list of SOUP items is maintained in GitHub in the Conda configuration files (`environment.yml`).
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply docs/source/software_design_overview.md:83]: I've removed excessive use of SOUP. Locking is coming soon, and the statement will still be true once that is in.

**Comment by peterhessey** [comment]
After received feedback from Addenbrooke's on this, it would be worth adding a more detailed description to the finalization section on the software engineering practices that are followed to review each new PR (reviewing code quality, unit tests, system tests, etc.).

File: docs/source/software_design_overview.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+The design and development of the software in this repository is roughly separated into an **initiation**,
+**prototyping**, and a **finalization** phase. The initiation phase can be skipped for minor changes, for example an
+update to documentation.
+
+### Initiation
+
+During the initiation phase, the following steps are carried out:
+
+- Collection of a set of requirements
```
Comment (peterhessey) [inline docs/source/software_design_overview.md:25]:
```suggestion
- Collection of a set of requirements.
```

File: docs/source/software_design_overview.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+- Creating a suggested design for the change
```
Comment (peterhessey) [inline docs/source/software_design_overview.md:26]:
```suggestion
- Creating a suggested design for the change.
```

File: docs/source/software_design_overview.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+- Review of the design with member of the core team
```
Comment (peterhessey) [inline docs/source/software_design_overview.md:27]:
```suggestion
- Review of the design with member of the core team.
```

File: docs/source/software_design_overview.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+The deliverables of this phase are a detailed design of the proposed change in a GitHub Issue or a separate document.
+
+### Prototyping
+
+The engineering owner of the proposed change will create a branch of the current codebase. This branch is separate from
+the released (main) branch to not affect any current functionality. In this branch, the engineer will carry out the
+changes as proposed in the design document.
+
+The engineer will also add additional software tests at different levels (unit tests, integration tests) as appropriate
+for the design. These tests ensure that the proposed functionality will be maintained also in future.
```
Comment (peterhessey) [inline docs/source/software_design_overview.md:38]:
```suggestion
for the design. These tests ensure that the proposed functionality will be maintained in future.
```

File: docs/source/software_design_overview.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+### Finalization
+
+At this point, the engineering owner of the proposed change has carried out all necessary changes in a branch of the
+codebase. They would now initiate a Pull Request process that consists of the following steps:
+
+- The code will be reviewed by at least 2 engineers. Both need to provide their explicit approval before the proposed
+  change can be integrated in the "main" branch.
+- All unit and integration tests will start.
+- All automatic code checks will start. These checks will verify
```
Comment (peterhessey) [inline docs/source/software_design_overview.md:51]:
```suggestion
- All automatic code checks will start. These checks will verify the following:
```

File: docs/source/software_design_overview.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+  - Consistency with static typing rules
```
Comment (peterhessey) [inline docs/source/software_design_overview.md:53]:
```suggestion
  - Consistency with static typing rules.
```

File: docs/source/software_design_overview.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+  - Ensure that no passwords or other types of credentials are revealed
```
Comment (peterhessey) [inline docs/source/software_design_overview.md:54]:
```suggestion
  - Ensure that no passwords or other types of credentials are revealed.
```

File: docs/source/software_design_overview.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+## Software Configuration Management
+
+Third party libraries (sometimes called Software of Unknown Provenance, SOUP) are consumed via two
+package management systems:
+
+- Conda
+- PyPi
+
+Both of those package management systems maintain strict versioning: Once a version of a package is published, it
```
Comment (peterhessey) [inline docs/source/software_design_overview.md:77]:
```suggestion
Both of those package management systems maintain strict versioning: once a version of a package is published, it
```

File: docs/source/software_design_overview.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+## Defect Handling
+
+The handling of any bugs or defects discovered is done via GitHub Issues.
+
+When an Issue is created, the team members will get notified. Open Issues are kept in a list sorted by priority. For
+example, [this is the list for all of the InnerEye-related Issues](https://github.com/orgs/microsoft/projects/320).
+Priorities are re-assesed regularly, at least once a week.
+
+The Issue(s) with highest priority are assigned to an engineer. The engineer would then analyze the problem, and
```
Comment (peterhessey) [inline docs/source/software_design_overview.md:98]:
```suggestion
The Issue(s) with highest priority are assigned to an engineer. The engineer will then analyze the problem, and
```

File: docs/source/software_design_overview.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+possibly request more information to reproduce the problem. Requesting more information is also handled in the GitHub
+Issue. Once the problem is clearly reproduced, the engineer will start to write a fix for the Issue, and a test that
```
Comment (peterhessey) [inline docs/source/software_design_overview.md:100]:
```suggestion
Issue. Once the problem is clearly reproduced, the engineer will start to write a fix for the Issue as well as a test that
```

**Review by peterhessey — APPROVED** [review]
Approved with minor comments

**Review by hamurfet — APPROVED** [review]

**Pull request merged**